### ISSUE TITLE
HMR randomly fails in browser console with no obvious cause/pattern

### ISSUE THREAD

**Issue by AaronBeaudoin** [remix-run/remix#7604]
### What version of Remix are you using?

2.0.1

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

None. I haven't been able to reliably reproduce this issue because it just happens randomly. Here's what I've been doing:

1. Link to a CSS file in `root.jsx` like so:

```js
import styles from "./styles.css";

export const links = () => [
  { rel: "stylesheet", href: styles }
];
```

2. Toggle a color back and forth between some values:

```css
.color {
  background-color: hsl(0, 100%, 50%);
}
```

3. After a few times the HMR will usually randomly fail with an error like this:

<img width="620" alt="image" src="https://github.com/remix-run/remix/assets/26592486/d666d896-918c-43b6-ae3c-c88dd10c2468">

### Expected Behavior

HMR either **always works** or **always doesn't** depending on whether my code is correct.

### Actual Behavior

HMR randomly sometimes doesn't work with no obvious reason.

I'm on MacOS 13.5.2, using Node 18.18.0, and Remix 2.0.1.

**Comment by fwardcorpaxe** [comment]
I've noticed the same thing in Ubuntu (through Windows WSL2).

It's fairly inconsistent but still seems to be race condition related.

**Comment by AaronBeaudoin** [comment]
I created this issue in response to a conversation in the Discord server. User @xHomu suggested I try creating [this](https://github.com/remix-run/remix/blob/main/templates/express/server.js) `server.js` file and see if I can create errors again while running `npx remix dev --manual -c "node server.js"`. To my surprise, I wasn't able to get the errors to come up. I went back and tried regular `npx remix dev` and I was able to get the errors again.

The way that I got the errors to come up was to create a quick script to toggle the CSS color really fast to try and bog down the server. Here is the simple code, which I ran with `node toggler.js` with the dev server already running:

```js
import fs from "fs";
let toggle = false;

setInterval(() => {
    fs.writeFileSync("app/color.css", `.color {
        background-color: hsl(${toggle ? 0 : 180}, 100%, 50%);
    }`);
    toggle = !toggle;
}, 200);
```

This class was applied in a simple `app/routes/_index.jsx` like this:

```jsx
export default function App() {
  return (
    <div className="color">
      Hello world!
    </div>
  );
}
```

And my `app/root.jsx` looks like this:

```jsx
import {
  Links,
  LiveReload,
  Meta,
  Outlet,
  Scripts,
} from "@remix-run/react";

import styles from "./color.css";

export const links = () => {
  return [
    {
      rel: "stylesheet",
      href: styles
    }
  ];
};

export default function App() {
  return (
    <html>
      <head>
        <link
          rel="icon"
          href="data:image/x-icon;base64,AA"
        />
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```

**Comment by xHomu** [comment]
That's odd, @pcattori shouldn't remix app server be the exact same as the Express template server.js?

**Comment by AaronBeaudoin** [comment]
Is there any way that [this](https://github.com/remix-run/remix/blob/main/packages/remix-serve/cli.ts#L48C1-L52C8) code could be making the difference?

**Comment by AaronBeaudoin** [comment]
Also, just a note that you'll need to change the interview in my `toggler.js` snippet above in order to "reliably" reproduce this issue. I found that on my Windows machine 200ms was a good rate, but on my Mac I had to lengthen the time to 350ms to get the error to show up most frequently.

**Comment by AaronBeaudoin** [comment]
_Nice_ vs. _not nice_ and "snark" wasn't a factor in my question, although I could have worded it better. I directly indicated the impression I got from reading through the Remix discussion on Vite and asked whether it was true.

It sounds like you're implying that Remix **_has_** invested significant time in DX. That's great to hear, although you also notably didn't address why Remix has had Vite on the radar for quite a while with no indication of what's next. DX obviously does matter to an extent in the Remix design—although transparency and alignment with web standards appears to be stronger defining goals. **There's nothing wrong with any of that.**

**Comment by brophdawg11** [comment]
I believe this is related to https://github.com/remix-run/remix/issues/7466, specifically [this comment](https://github.com/remix-run/remix/issues/7466#issuecomment-1726445564)

**Comment by AaronBeaudoin** [comment]
That [comment](https://github.com/remix-run/remix/issues/7466#issuecomment-1726445564) from @pcattori seems to check out. Yet I was experiencing the issue without fail for **every single HMR update**. (I only put together the script above in an attempt to make it more reproducible.) So under certain conditions it is clearly possible for the issue to occur pretty much without fail.

I suspect that in my case this is the problem:

1. I wanted to use [UnoCSS](https://unocss.dev) (it's a Tailwind compatible CSS engine that totally kicks butt), so I installed these dependencies:

   ```json
   "devDependencies": {
     "unocss": "^0.56.5",
     "concurrently": "^8.2.1"
   }
   ```

2. I can't use the Vite integration for UnoCSS because Remix doesn't use Vite, and the PostCSS integration for UnoCSS is currently experimental. So I added scripts like this to simply use the UnoCSS CLI in watch mode, which from my research seems to align with the "Remix" approach for situations like this:

   ```jsonc
   "scripts": {
     // 1. Watch files matching app/**/*.{js,jsx,ts,tsx}
     // 2. Output classes into app/uno.css
     "unocss:dev": "unocss -wo 'app/uno.css' 'app/**/*.{js,jsx,ts,tsx}'",
     // 3. Run UnoCSS concurrently with `remix dev`
     "dev": "conc -n R,U 'remix dev' npm:unocss:dev"
   }
   ```

3. Now I simply include `app/uno.css` in `root.jsx` like this (the other import is just the base CSS reset which UnoCSS does not include in the output for flexibility reasons):

   ```js
   import styleReset from "@unocss/reset/tailwind.css";
   import styleUno from "./uno.css";

   export const links = () => [
     { rel: "stylesheet", href: styleReset },
     { rel: "stylesheet", href: styleUno }
   ];
   ```

4. Finally, I run `npm run dev`. Whenever I make an update to CSS classes in my JSX it immediately triggers a rebuild. Since UnoCSS is in watch mode, it also picks up the change and updates `app/uno.css`, so my theory is that this is triggering a second rebuild so quickly that it's basically guaranteed the browser won't be finished with the first HMR update. I'm not sure how to verify this, though.

**Comment by akomm** [comment]
Yes, I think [this issue](https://github.com/remix-run/remix/issues/7466) is the same.

**Comment by brophdawg11** [comment]
Thank you for opening this issue, and our apologies we haven't gotten around to it yet!

With the release of [React Router v7](https://remix.run/blog/react-router-v7) we are sun-setting continued development/maintenance on Remix v2. If you have not already [upgraded to React Router v7](https://reactrouter.com/upgrading/remix), we recommend you do so. We've tried to make the upgrade process as smooth as possible with our [Future Flags](https://remix.run/docs/en/main/start/future-flags). We are now in the process of cleaning up outdated issues and pull requests to improve the overall hygiene of our repositories.

We plan to continue to address 2 types of issues in Remix v2:

- Bugs that pose security concerns
- Bugs that prevent upgrading to React Router v7

If you believe this issue meets one of those criteria, please respond or [create a new issue](https://github.com/remix-run/remix/issues/new?template=bug_report.yml).

For all other issues, ongoing maintenance will be happening in React Router v7, so:

- If this is a bug, please reopen this issue [in that repo](https://github.com/remix-run/react-router/issues/new?template=bug_report.yml) with a new [minimal reproduction](https://reactrouter.com/new) against v7
- If this is a feature request, please open a new [Proposal Discussion](https://github.com/remix-run/react-router/discussions/new?category=proposals) in React Router, and if it gets enough community support it can be considered for implementation

If you have any questions you can always reach out on [Discord](https://rmx.as/discord). Thanks again for providing feedback and helping us make our framework even better!
